### DISCUSSION TITLE
Why return File data undefined when I upload file type input?

### DISCUSSION THREAD

**Discussion by Imjurney** [react-hook-form/react-hook-form#9869] [Q&A] [answered]
Hi! PLZ help me. I can't find solution.
I'm korean, so I am not good at English. Thank you for your understanding🙏
I am now managing two type input components separately.
In the standalone component state, the file is saved just fine.
But when I combine them and send them to the form, undefined return...
<img width="454" alt="image" src="https://user-images.githubusercontent.com/102623672/216801322-a6fcc10c-ed91-40f3-974c-d438892eee1e.png">
here's my code.

```
import { CareerInput } from '@components/CareerInput';
import { NoFooterLayout } from '@components/Layout/NoFooterLayout';
import { supabase } from '@lib/supabase';
import clsx from 'clsx';
import { FieldValues, FormProvider, useForm } from 'react-hook-form';
import styles from '@components/CareerInput/CareerInput.module.css';
import { Button, ProfileUploader, TextInput } from '@components/index';
import router from 'next/router';
import { ImageInputButton } from '@components/ImageInputList/ImageInputButton/ImageInputButton';
import {
  ChangeEvent,
  ReactNode,
  SetStateAction,
  useEffect,
  useState,
} from 'react';


export default function AddUserInfo(props: ReactNode) {
  const {
    register,
    handleSubmit,
    watch,
    control,
    formState: { errors },
  } = useForm<FieldValues>({
    mode: 'onBlur',
  });

  const [image, setImage] = useState({
    image_file: '',
    preview_URL: '/assets/default_profile.png',
  });

  const onsumit = (Formdata: FieldValues) => {
    console.log(Formdata);

  const saveImage = (e: ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    if (e.target.files !== null && e.target.files[0]) {
      // 새로운 이미지를 올리면 createObjectURL()을 통해 생성한 기존 URL을 폐기
      if (e.target.files[0].size > 2097152) {
        return alert('파일이 너무 큽니다');
      }
      URL.revokeObjectURL(image.preview_URL);
      const preview_URL = URL.createObjectURL(e.target.files[0]);
      setImage(() => ({
        image_file: (e.target.files as FileList)[0] as unknown as string,
        preview_URL: preview_URL,
      }));
    }
  };

  return (
    <NoFooterLayout>
      <form onSubmit={handleSubmit(onsumit)}>
        <h2 className="font-semibold text-2xl mb-8">추가 정보</h2>
        <fieldset className="flex flex-col justify-center items-center gap-20">
          <legend className="sr-only">추가 정보 입력</legend>
          <ProfileUploader register={register} /> // **<-- problem**
          <div className="flex gap-10 items-center primaryFocusVisible">
            <h2 className="text-sm">운동 경력이 얼마나 되시나요? *</h2>
            <CareerInput
              register={register}
              id="career"
              className={clsx(styles.inputCommon, styles.inputWideSize)}
              formName="career"
              selectName="career"
            />
          </div>
          <Button type="submit" className="w-full h-10 mt-20">
            <p className="text-base">추가 정보 업데이트</p>
          </Button>
        </fieldset>
      </form>
    </NoFooterLayout>
  );
}

```

```
import { ChangeEvent, useEffect, useState } from 'react';
import { Button } from '@components/Button';
import Image from 'next/image';
import clsx from 'clsx';
import { EditButton } from '@components/index';
import { HiOutlineCamera } from 'react-icons/all';
import { UseFormRegister, FieldValues } from 'react-hook-form';

interface ProfileProps {
  size?: 'small' | 'medium' | 'large';
  name?: string;
  register: UseFormRegister<FieldValues>;
}

const sizes = {
  small: 'w-24 h-24',
  medium: 'w-40 h-40',
  large: 'w-52 h-52',
};

export const ProfileUploader = ({
  size = 'small',
  name = 'image_file',
  register,
  ...props
}: ProfileProps) => {
  const [image, setImage] = useState({
    image_file: '',
    preview_URL: '/assets/default_profile.png',
  });
  console.log(image);
  let inputRef: HTMLInputElement | null;

  const saveImage = (e: ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    if (e.target.files !== null && e.target.files[0]) {
      // 새로운 이미지를 올리면 createObjectURL()을 통해 생성한 기존 URL을 폐기
      if (e.target.files[0].size > 2097152) {
        return alert('파일이 너무 큽니다');
      }
      URL.revokeObjectURL(image.preview_URL);
      const preview_URL = URL.createObjectURL(e.target.files[0]);
      setImage(() => ({
        image_file: (e.target.files as FileList)[0] as unknown as string,
        preview_URL: preview_URL,
      }));
    }
  };

  const deleteImage = () => {
    // createObjectURL()을 통해 생성한 기존 URL을 폐기
    URL.revokeObjectURL(image.preview_URL);
    setImage({
      image_file: '/',
      preview_URL: '/assets/default_profile.png',
    });
  };

  useEffect(() => {
    // 컴포넌트가 언마운트되면 createObjectURL()을 통해 생성한 기존 URL을 폐기
    return () => {
      URL.revokeObjectURL(image.preview_URL);
    };
  }, [image]);
  return (
    <>
      <div className={clsx('relative', sizes[size])}>
        <EditButton
          className="bg-orange-400 w-6 h-6 border border-neutral-100 bottom-0 right-1 rounded-full absolute"
          icon={<HiOutlineCamera className="stroke-neutral-100 ml-[0.2rem]" />}
          onClick={() => inputRef?.click()}
          title="나가기"
          type="button"
        />
        <input
          className="hidden"
          type="file"
          accept="image/jpg, image/jpeg"
          // 클릭할 때 마다 file input의 value를 초기화 하지 않으면 버그가 발생할 수 있다
          // 사진 등록을 두개 띄우고 첫번째에 사진을 올리고 지우고 두번째에 같은 사진을 올리면 그 값이 남아있음!
          onClick={(e) => {
            const target = e.target as HTMLInputElement;
            target.value = '';
          }}
          {...(register &&
            register(name, {
              onChange: (e: ChangeEvent<HTMLInputElement>) => {
                saveImage(e);
              },
            }))}
          ref={(refParam) => (inputRef = refParam)}
          {...props}
        />
        <div className=" object-contain flex flex-col gap-2 items-center">
          <Image
            priority
            onClick={() => inputRef?.click()}
            className={clsx(
              'rounded-full object-contain self-center',
              sizes[size]
            )}
            src={image.preview_URL}
            alt={''}
            width={100}
            height={100}
          />

          <Button
            color="Outline"
            className="p-1"
            size="xxs"
            onClick={deleteImage}
          >
            <p>삭제</p>
          </Button>
        </div>
      </div>
    </>
  );
};

```

**Comment by leapful** [comment, marked as answer]
@Imjurney 

Could you please provide a codesandbox url to re-produce the issue ?

**Reply by Imjurney** [reply]
@leapful  hi! 😊 
I moved them as you said.
But there's no console.log(Formdata) return on the code sandbox. 😭 
Can you take a look?
https://codesandbox.io/s/polished-forest-tc8t5d?file=/public/index.html

thanks!!

**Reply by leapful** [reply]
@Imjurney 

The problem on "no console.log" from codesanbox url is because you are using Button component with `type=button` therefore it cannot be trigger submit event on form element, it should be `type=submit`.

The main problem of undefined on image File data is because react-hook-form only supports basic primitive types when serializing data to `handleSubmit`.

What is your purpose of using File type when submitting the form ? If it's mainly because you can to submit the content of the file image to server-side, you can transform the binary content to base64 then set to form values.

**Reply by Imjurney** [reply]
@leapful 

I really appreciate your help!

the feature is for update when new user signup!

So the purpose is to submit the content of the file image to server-side just like you said!

Could you teach me guideline?

**Reply by leapful** [reply]
@Imjurney 

You need to get the target File object via `onChange` event handler of your `<input />` element then transform it to data url base64. After that, just trigger `onChange` callback from method `register` to update the field "image_file".

Just remember to ask your back-end team to transform back field "image_file" from data url to binary content if needed.

Working demo: https://codesandbox.io/s/adoring-booth-7z03ue?file=/src/FileUploader.tsx

**Reply by Imjurney** [reply]
@leapful wow!! Thank you very much!! 
It helped me a lot!! 😭👍

**Comment by fbolcic** [comment]
Hello @Imjurney!

We are encountering the same problem here and I would like to hear your opinion about something. 

Basically, we have a controlled `FileField` that manages an uploaded `File` under the hood. The thing is, we were using this `FileField` with `react-hook-form` and had no issues up until recently, when we integrated it in a slightly more complex form. 
By no issues I mean that the field was able to preserve its `File` object, and the form library provided it in our submit handler afterwards.

Like I said, the difference between the two forms is in their complexity.
Form 1 (less complex) uses `react-hook-form` features:
1. Controlled fields
2. Submit handler
3. isDirty, isValid, isSubmitting flags

Form 2 (more complex) uses the same features and also:
1. dirtyFields 
2. getValues
3. watch, useWatch

So in Form 1 we have no issue with the field, the uploaded `File` is provided all the way to the submit handler. 
In Form 2 the issue arises, and the `File` becomes `undefined` somewhere along the way to the submit handler. This obviously causes issues for us since we cannot then upload it to the BE.

But since Form 2 uses a bit more of the `hook-form` API, it seems to me that one of this APIs mutates the field and causes the `File` attribute to become `undefined`. 
Like I said, this form is significantly more complex and it is hard for us to figure out which API is the culprit. The complexity also makes it hard to reproduce in a sandbox. 

Do you have an idea where could this mutation occur? It would help us decide how to approach this problem and how to optimize our features.

Any help would be greatly appreciated!

**Reply by Imjurney** [reply]
Hello.
Is it correct that the undefined appears when the Filedata is submitted to the handlesubmit and handed over to BE? I'm not good at English because I'm Korean.
